src/server/project.ts

```diff
@@ -1520,6 +1540,53 @@ namespace ts.server {
             this.canonicalConfigFilePath = asNormalizedPath(projectService.toCanonicalFileName(configFileName));
         }
 
+        /* @internal */
+        setResolvedProjectReferenceCallbacks(projectReferenceCallbacks: ResolvedProjectReferenceCallbacks) {
+            this.projectReferenceCallbacks = projectReferenceCallbacks;
+        }
+
+        /* @internal */
+        useSourceOfProjectReferenceRedirect = () => !!this.languageServiceEnabled &&
+            !this.getCompilerOptions().disableSourceOfProjectReferenceRedirect;
+
+        fileExists(file: string): boolean {
+            // Project references go to source file instead of .d.ts file
+            if (this.useSourceOfProjectReferenceRedirect() && this.projectReferenceCallbacks) {
+                const source = this.projectReferenceCallbacks.getSourceOfProjectReferenceRedirect(file);

```

Comment:
It feels a little scary that these (public API) methods might do something fairly different from what the name implies. Are there a lot of callers to them or is it mostly the ProjectService? Since it’s publicly exposed, maybe at least add a note in the JSDoc that it might un-redirect declaration files?

Reply:
This is public method since this is method for LanguageServiceHost and should result in answer `yes file exists` for .d.ts files even if they are not there to ensure that we can use source file instead..  Will add comment